#### **Welcome to Assignment 3 on Deep Learning for Computer Vision.**
This assignment is based on the content you learned in Week-3.

#### **Instructions**
1. Use Python 3.x to run this notebook
2. Write your code only in between the lines 'YOUR CODE STARTS HERE' and 'YOUR CODE ENDS HERE'. You should not change anything else in the code cells, if you do, the answers you are supposed to get at the end of this assignment might be wrong.
3. Read documentation of each function carefully.
4. All the Best!

# Pytorch operations

You will perform basic elementwise operations between two matrices

Step1: rearrange the give input x into a tensor of shape (10,5,3) and store it in variable x_resize \

Step2: Generate a gaussian random tensor (mean =0, variance=1) of shape (5,3) and store it in variable y1 \
Step3:  Generate a uniform random tensor (interval [-10,10)) of shape (5,3) and store it in variable y2 \

Step4: Perform elementwise multiplication between x_resize and y1, store it in variable mul_output1 \

Step5: Find the max element along the last dimension of mul_output1 and store its value in final_output1 \

Step6: Repeat Step4 and Step5 for x_resize and y2 and store the corresponding values in the variables, mul_output2 and final_output2.

In [1]:
import torch

In [2]:
import torch
#DO NOT CHANGE THE SEED
torch.manual_seed(2)

x = torch.randint(low=0,high=256,size=(1,150))
### YOUR CODE STARTS HERE ###

x_resize = x.view(10, 5, 3)

# Use torch.randn()
y1 = torch.randn(5, 3)

# Use torch.rand()
y2 = torch.rand(5, 3) * 20 - 10

mul_output1 = x_resize * y1

final_output1,_ = torch.max(mul_output1, dim=2)

mul_output2 = x_resize * y2

final_output2,_ = torch.max(mul_output2, dim=2)

### YOUR CODE ENDS HERE ###

print(torch.mean(final_output1+final_output2))

tensor(879.7391)


### Question 1. What is mean of the sum of final_output1 and final_output2? (Select the closest value)

a) 875

b) 880

c) 883

d) 870


# Custom loss function and its Gradients

In this we will implement a custom loss function and calculate the gradient of this loss function w.r.t two input tensors.

The loss function takes two inputs, y_true (given) and y_pred. Follow the below step to generate y_pred.

1) Calculate the matrix-vector multiplication between 'M' and 'v' and store it in the variable "y_pred".


In [9]:
import torch
import torch.nn as nn


#DO NOT CHANGE THE SEED
torch.manual_seed(0)

y_true = torch.tensor([0.3, 0.7, 0.6])

m, n = 3, 2
M = torch.randn(m, n, requires_grad=True)
v = torch.randn(n, requires_grad=True)

### YOUR CODE STARTS HERE ###
y_pred = torch.matmul(M, v)
### YOUR CODE ENDS HERE ###

Next, implement a custom loss function in PyTorch that computes the mean squared error (MSE) between two tensors (y_true, y_pred), but only for elements where the target tensor y_true is greater than 0.5. Save the calculated MSE loss in a variable named "loss".

*(Hint: First create a mask on y_true and then multiply the mask with the loss)*

In [10]:
import torch
import torch.nn as nn


#DO NOT CHANGE THE SEED
torch.manual_seed(0)

class CustomMSELoss(nn.Module):
    def __init__(self):
        super(CustomMSELoss, self).__init__()

    def forward(self, y_pred, y_true):
        ### YOUR CODE STARTS HERE ###
        mask = y_true > 0.5
        squared_diff = (y_pred - y_true) ** 2
        masked_squared_diff = squared_diff * mask 

        # For calculating the mean, take the sum of masked_squared_diff and divide by 3.
        loss = masked_squared_diff.sum() / 3
        ### YOUR CODE ENDS HERE ###
        return loss

### DO NOT CHANGE ###
loss_fn = CustomMSELoss()
loss = loss_fn(y_pred, y_true)
print("Loss: ", loss)

Loss:  tensor(2.0324, grad_fn=<DivBackward0>)


Next, calculate the gradient of the loss w.r.t tensors M and v.

Store the gradient w.r.t M in a variable grad_M

Store the gradient w.r.t v in a variable grad_v

(Hint: Look into the autograd library of Pytorch. https://pytorch.org/docs/stable/generated/torch.autograd.grad.html)


In [5]:
### DO NOT CHANGE ###
# Compute gradients using autograd
loss.backward()  # This computes gradients for all tensors that have requires_grad=True

### YOUR CODE STARTS HERE ###
grad_M = M.grad
grad_v = v.grad
### YOUR CODE ENDS HERE ###

print("Sum of Gradient Means: ", grad_M.mean()+grad_v.mean())

Sum of Gradient Means:  tensor(1.9638)


c:\Users\spars\AppData\Local\Programs\Python\Python312\Lib\site-packages\torch\autograd\graph.py:768: UserWarning: CUDA initialization: Unexpected error from cudaGetDeviceCount(). Did you run some cuda functions before calling NumCudaDevices() that might have already set an error? Error 1: invalid argument (Triggered internally at C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\c10\cuda\CUDAFunctions.cpp:108.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass


### Question 2: What is the loss value calculated using the Custom Loss function? (Select the closest value)

a) 1

b) 2.5

c)1.5

d) 2



### Question 3: What is sum of the means of the gradients, grad_M and grad_v? (Select the closest value)

a) 2

b) 1.7

c) 2.5

d) 3



# Multi Layer Perceptron

In this, we will implement a simple multilayer feedforward neural network using PyTorch with the following specifications:

Input dimension: 10

Hidden layer 1 dimension: 20

Hidden layer 2 dimension:  50

Output dimension: 5

Activation function: ReLU for the hidden layer

Your function should take a tensor x of shape (batch_size, 10) as input and return the output of the neural network.

In [ ]:
import torch
import torch.nn as nn

#DO NOT CHANGE THE SEED
torch.manual_seed(2)

# Define the multi-layer neural network class
class MultiLayerNN(nn.Module):
    def __init__(self):
        super(MultiLayerNN, self).__init__()

        ### YOUR CODE STARTS HERE ###
        self.fc1 = ___ # Input layer to first hidden layer
        self.fc2 = ___  # First hidden layer to second hidden layer
        self.fc3 = ___   # Second hidden layer to output layer
        ### YOUR CODE ENDS HERE ###

    def forward(self, x):
       ### YOUR CODE STARTS HERE ###
        x = ___
        x = ___

        ### YOUR CODE ENDS HERE ###
        x = self.fc3(x)
        return x

# DO NOT CHANGE:
model = MultiLayerNN()
input_tensor = torch.randn(32, 10)
output = model(input_tensor)
print("Mean of the output: ",output.mean()*1000)

### Question 4: What is mean of the output? (Select the closest value)

a) 1.6

b) 2.2

c) 3

d) 1.8



#Training of Multilayer Perceptron on a random dataset

In this we will use the Mulitlayer Perceptron defined in question 4 and train it on a random dataset.

In [ ]:
# Please DO NOT modify this cell.

import os
import os.path as osp
import random

import numpy as np
import torch

def set_seed(seed: int):

    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    if torch.cuda.is_available():
        torch.cuda.manual_seed(seed)
        torch.cuda.manual_seed_all(seed)
        torch.backends.cudnn.deterministic = True
        torch.backends.cudnn.benchmark = False
    os.environ["PYTHONHASHSEED"] = str(seed)

In [ ]:
# Please DO NOT modify this cell.

num_features = 10
classes = [0, 1, 2, 3, 4]
num_classes = len(classes)

num_samples = 100
num_train = 70
num_test = num_samples - num_train

In [ ]:
# Please DO NOT modify this cell.
# We are creating a random feature set and a random label set.
# The features and labels have no semantic meaning and might as well be garbage.

set_seed(2022)

features = np.random.random_sample((num_samples, num_features))
labels = np.random.choice(classes, size = num_samples)

# Train-test split
x_train = features[:num_train]
x_test = features[num_train:num_samples]

x_train = torch.Tensor(x_train)
x_test = torch.Tensor(x_test)

y_train = labels[:num_train]
y_test = labels[num_train:num_samples]

y_train = torch.LongTensor(y_train)
y_test = torch.LongTensor(y_test)

In [ ]:
# Please DO NOT modify this cell.

print(f"Train features: {x_train.shape}")
print(f"Test features: {x_test.shape}")

print(f"Train labels: {y_train.shape}")
print(f"Train labels: {y_test.shape}")


### Create train and test TensorDatasets from the respective numpy arrays

In [ ]:
# Create train and test TensorDatasets from the respective numpy arrays

#### YOUR CODE STARTS HERE ####

train_dataset = ___
test_dataset = ___

#### YOUR CODE ENDS HERE ####

### Create dataloaders using the datasets created in the previous cell.

In [ ]:
# Create dataloaders using the datasets created in the previous cell.
# Use a batch size of 64

#### YOUR CODE STARTS HERE ####

batch_size = ___

train_loader = ___
test_loader = ___

#### YOUR CODE ENDS HERE ####

### Training function

In [ ]:
def train(model, device, train_loader, optimizer, criterion, epoch):
    model.train()
    epoch_loss = 0
    for batch_idx, (data, target) in enumerate(train_loader):

        #### YOUR CODE STARTS HERE ####

        # send the data, target to the device
        data = ___
        target = ___

        # flush out the gradients stored in optimizer
        ___

        # pass the batch to the model and assign the output to variable named y_pred
        output = ___

        # calculate the loss (use CrossEntropyLoss in pytorch)
        loss = ___

        # do a backward pass
        ___

        # update the weights
        ___

        #### YOUR CODE ENDS HERE ####

        # Store loss
        epoch_loss += loss.item() * data.shape[0]

    print(f"Train Average Loss: {epoch_loss/len(train_loader.dataset):.2f}")

### Test function

In [ ]:
def test(model, device, test_loader, criterion, mode):
    model.eval()
    test_loss = 0
    correct = 0

    with torch.no_grad():
        for data, target in test_loader:
            #### YOUR CODE STARTS HERE ####

            # send data, target to the device
            data = ___
            target = ___

            # pass the image to the model and assign the output to variable named output
            output = ___
            #### YOUR CODE ENDS HERE ####

            test_loss += criterion(output, target).item() * data.shape[0]  # sum up batch loss
            pred = output.argmax(dim = 1, keepdim = True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)
    test_acc = 100. * correct / len(test_loader.dataset)

    print(f"{mode} Average loss: {test_loss:.2f}")
    print(f"{mode} Accuracy: {correct}/{len(test_loader.dataset)} ({test_acc:.2f}%)")

In [ ]:
set_seed(2022)

num_epochs = 300

#### YOUR CODE STARTS HERE ####

# check availability of GPU and set the device accordingly
device = ___

# Initialize MLP model
model = ___

# Define Adam Optimizer with a learning rate of 0.001
optimizer = ___

# Define CrossEntropyLoss as the criterion
criterion = ___

#### YOUR CODE ENDS HERE ####

for epoch in range(1, num_epochs+1):
    print(f"\nEpoch: {epoch}/{num_epochs}")

    train(model, device, train_loader, optimizer, criterion, epoch)
    test(model, device, test_loader, criterion, mode = "Test")

In [ ]:
test(model, device, train_loader, criterion, mode = "Train")

### Question 5

Report the final train accuracy (If you are not getting the exact number shown in
options, please report the closest number).

a. 100%

b. 53%

c. 61%

d. 79%



In [ ]:
test(model, device, test_loader, criterion, mode = "Test")

### Question 6

Report the final test accuracy (If you are not getting the exact number shown in
options, please report the closest number).

a. 30%

b. 40%

c. 15%

d. 25%



In [1]:
from collections import Counter

def findFirstUnique(s):
    # Count the frequency of each character
    frequency = Counter(s)
    
    # Iterate through the string to find the first unique character
    for index, char in enumerate(s):
        if frequency[char] == 1:
            return index + 1  # 1-based index

    # If no unique character is found, return -1
    return -1

# Example usage
s = "statistics"
result = findFirstUnique(s)
print(result)  # Output should be 3


3


In [2]:
from collections import Counter

def findFirstUnique(s):
    # Count the frequency of each character
    frequency = Counter(s)
    
    # Iterate through the string to find the first unique character
    for index, char in enumerate(s):
        if frequency[char] == 1:
            return index + 1  # 1-based index

    # If no unique character is found, return -1
    return -1

# Example usage
s = "hackthegame"
result = findFirstUnique(s)
print(result)  # Output should be the index of the first unique character

3


In [2]:
l = [1,2,3,4]
print(l[2:3])

[3]


In [3]:
import numpy as np

def elu(x, alpha=0.3):
    return x if x > 0 else alpha * (np.exp(x) - 1)

def elu_tanh(x, alpha=0.3):
    tanh_x = np.tanh(x)
    return elu(tanh_x, alpha)

In [4]:
x = -1.3  # Replace with any value of x
alpha = 0.3  # Replace with the desired alpha value

result = elu_tanh(x, alpha)
print(f"ELU(tanh({x})) = {result:.2f}")

ELU(tanh(-1.3)) = -0.17
